modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteServerImpl.java

```diff
@@ -82,10 +86,33 @@ public class IgniteServerImpl implements IgniteServer {
 
     private final ClassLoader classLoader;
 
-    private volatile @Nullable IgniteImpl instance;
+    private final Executor asyncContinuationExecutor;
+
+    /** Current Ignite instance. This field is not volatile to make hot path accesses from IgniteReference and other references
+     * faster (they always happen under a read lock, which guarantees visibility of changes to this field). So we access
+     * this field in this object under synchronization ({@link #igniteMonitor} serves as the monitor).
+     */
+    private @Nullable IgniteImpl ignite;
+
+    private final Object igniteMonitor = new Object();
```

Comment:
I thought the classic name for such objects was "igniteMux".

Reply:
Renamed it to `Mutex`. `Mux` is funny, but it's a mutex, after all :)

Reply:
Ok